File: py/src/pypaginate/query.py
"""One-shot filter / sort / search over in-memory sequences.

The ergonomic complement to :func:`pypaginate.paginate`: filter, sort, or search
a list directly through the native engine. Each returns a new list of matching
host items (search in ranked relevance order).
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import TypeVar

from pypaginate import _native
from pypaginate.specs import FilterGroup, FilterSpec, SearchSpec, SortSpec


ItemT = TypeVar("ItemT")

#: A single spec, a sequence of them (each keeps its own logic), or a group.
FilterWhere = FilterSpec | Sequence[FilterSpec] | FilterGroup
#: A single sort key or a sequence applied in priority order.
SortBy = SortSpec | Sequence[SortSpec]


def filter(items: Sequence[ItemT], where: FilterWhere) -> list[ItemT]:  # noqa: A001
    """Return the items matching ``where``, in original order."""
    rows = list(items)
    if isinstance(where, FilterGroup):
        indices = _native.filter_group_indices(rows, where)
    else:
        specs = [where] if isinstance(where, FilterSpec) else list(where)
        indices = _native.filter_indices(rows, specs)
    return [rows[i] for i in indices]


def sort(items: Sequence[ItemT], by: SortBy) -> list[ItemT]:
    """Return the items ordered by ``by`` (stable, null-aware)."""
    rows = list(items)
    specs = [by] if isinstance(by, SortSpec) else list(by)
    return [rows[i] for i in _native.sort_indices(rows, specs)]


def search(items: Sequence[ItemT], spec: SearchSpec) -> list[ItemT]:
    """Return the items matching ``spec`` in ranked (relevance) order."""
    rows = list(items)
    return [rows[i] for i in _native.search_indices(rows, spec)]


__all__ = ["FilterWhere", "SortBy", "filter", "search", "sort"]

File: ts/src/query.ts
/**
 * One-shot item-returning query helpers — the ergonomic complement to the
 * index-returning {@link filterIndices}/{@link sortIndices}/{@link searchIndices}
 * and the mirror of pypaginate's top-level `filter` / `sort` / `search`. Each
 * returns a new array of the matching items (search in ranked order).
 *
 *     import { search, filter, sort } from "@cyblow/paginate";
 *     const adults = filter(users, { field: "age", operator: "gte", value: 18 });
 *     const hits = search(users, { query: "alice", fields: ["name", "email"] });
 */
import { filterGroupIndices, filterIndices, searchIndices, sortIndices } from "./engines.js";

import type { FilterGroup, FilterSpec, SearchSpec, SortSpec } from "./specs.js";

/** A single spec, a flat list (combined per each `logic`), or a nested group. */
export type FilterWhere = FilterSpec | readonly FilterSpec[] | FilterGroup;
/** A single sort key or a sequence applied in priority order. */
export type SortBy = SortSpec | readonly SortSpec[];

function isGroup(where: FilterWhere): where is FilterGroup {
  return !Array.isArray(where) && "conditions" in where;
}

/** Return the items matching `where`, in original order. */
export function filter<T extends object>(items: readonly T[], where: FilterWhere): T[] {
  let indices: number[];
  if (isGroup(where)) {
    indices = filterGroupIndices(items, where);
  } else {
    const specs = Array.isArray(where) ? where : [where];
    indices = filterIndices(items, specs as readonly FilterSpec[]);
  }
  return indices.map((i) => items[i]);
}

/** Return the items ordered by `by` (null-aware, stable). */
export function sort<T extends object>(items: readonly T[], by: SortBy): T[] {
  const specs = (Array.isArray(by) ? by : [by]) as readonly SortSpec[];
  return sortIndices(items, specs).map((i) => items[i]);
}

/** Return the items matching `spec` in ranked (relevance) order. */
export function search<T extends object>(items: readonly T[], spec: SearchSpec): T[] {
  return searchIndices(items, spec).map((i) => items[i]);
}
